lib/widgets/path_editor/editors/graph_editor.dart

```diff
@@ -0,0 +1,427 @@
+
+  Widget leftTitleWidgets(double value, TitleMeta meta) {
+    if ((value - value.toInt()).abs() > 0.01) {
+      return const Text('');
+    }
+
+    const style = TextStyle(
+      color: Colors.white,
+      fontWeight: FontWeight.bold,
+      fontSize: 14,
+    );
+
+    return SideTitleWidget(
+        axisSide: meta.axisSide,
+        space: 10,
+        child: Text(value.toInt().toString(),
+            style: style, textAlign: TextAlign.center));
+  }
+
+  Widget bottomTitleWidgets(double value, TitleMeta meta) {
+    if (((value * 10.0) - (value * 10.0).round().toInt()).abs() > 0.01) {
+      return const Text('');
+    }
+
+    const style = TextStyle(
+      color: Colors.white,
+      fontWeight: FontWeight.bold,
+      fontSize: 16,
+    );
+
+    return SideTitleWidget(
+      axisSide: meta.axisSide,
+      space: 10,
+      child: Text(value.toStringAsFixed(1), style: style),
+    );
+  }
+}
+
+class GraphEditor extends StatefulWidget {
+  final RobotPath path;
+  final bool holonomicMode;
+  final void Function(RobotPath path) savePath;
+  final SharedPreferences prefs;
+  static String prefShowVelocity = 'showVelocity';
+  static String prefShowAccel = 'showAccel';
+  static String prefShowHeading = 'showHeading';
+  static String prefShowRotation = 'showRotation';
+  static String prefShowAngularVelocity = 'showAngularVelocity';
+  static String prefShowCurvature = 'showCurvature';
+  static Color colorVelocity = Colors.red;
+  static Color colorAccel = Colors.orange;
+  static Color colorHeading = Colors.yellow;
+  static Color colorRotation = Colors.yellow;
+  static Color colorAngularVelocity = Colors.green;
+  static Color colorCurvature = Colors.blue;
+
+  const GraphEditor(
+      {required this.path,
+      required this.holonomicMode,
+      required this.savePath,
+      required this.prefs,
+      super.key});
+
+  @override
+  State<GraphEditor> createState() => _GraphEditorState();
+}
+
+class _GraphEditorState extends State<GraphEditor> {
+  final GlobalKey _key = GlobalKey();
+  UniqueKey _graphRuntimeKey = UniqueKey();
+  bool isSampled = false;
+  bool cardMinimized = false;
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      key: _key,
+      children: [
+        Padding(
+          padding: const EdgeInsets.fromLTRB(48, 48, 48, 72),
+          child: _LineChart(
+              isSampled: isSampled,
+              path: widget.path,
+              holonomicMode: widget.holonomicMode,
+              prefs: widget.prefs),
+        ),
+        _buildGraphSettingsCard(),
+        _buildGeneratorSettingsCard(),
+      ],
+    );
+  }
+
+  Widget _buildGraphSettingsCard() {
+    return GraphSettingsCard(
+        stackKey: _key,
+        key: _graphRuntimeKey,
+        holonomicMode: widget.holonomicMode,
+        onUpdate: () async {
+          setState(() {
+            // Force rebuild card to update runtime
+            _graphRuntimeKey = UniqueKey();
+          });
+        },
```

Comment:
Not needed with changes to graph_settings_card

```suggestion
```

Reply:
The only thing that you are passing _graphRuntimeKey to that I can see is the GraphSettingsCard widget so I don't think this is actually causing the graph itself to rebuild. I'm not sure though.

Reply:
The graph isn't rebuilt without it! Not sure if it is because of the _graphRuntimeKey or the setState() by itself...I'm still a bit unclear on the Flutter event model.

I followed the pattern used on other cards/editors, but may have mis-applied it a bit. I got rid of _graphRuntimeKey and have an empty lambda in setState() and it redraws the graph. I'm happy with that if you are.

Reply:
I'm wondering if that's because of the GlobalKey you have. Not sure how that works. Does it still rebuild the graph if you use a UniqueKey for _key and use `_key = UniqueKey()` in set state? If that works it would be a bit more clear that rebuilding the graph is what its doing.

Reply:
Didn't notice that. It should be fine but what you could do instead is create a new field for the UniqueKey and only pass that to the _LineChart widget.

Reply:
Sounds reasonable. I'll look into that.

Reply:
So, the _LineChart widget is a StatelessWidget, meaning that the UniqueKey passed to it with this approach would not change (if I'm understanding Dart correctly, it is pass by value only, so it couldn't change; and the key in _LineChart would have to be final since it is a StatelessWidget, correct?).

In the end, I think this ends up just being syntactic sugar. This is an odd case in that the graph needs to be redrawn based on "state" stored in the preferences that is updated by the card...so there is no state update required in the GraphEditor or _LineChart widgets.

I'm leaning towards an empty lambda with a comment that describes what is happening. As opposed to trying to imply it with a meaningless "state" change and hoping that someone understands it a couple of years from now. Thoughts?